one/src/migrations.rs

```diff
@@ -84,34 +84,68 @@ async fn from_ipfs(opts: FromIpfsOpts) -> Result<()> {
     let network = opts.network.to_network(&opts.local_network_id)?;
     let db_opts: DBOpts = (&opts).into();
     let crate::Databases::Sqlite(db) = db_opts.get_database().await?;
-    let blocks = blocks_from_filesystem(opts.input_ipfs_path);
+    let blocks = FSBlockStore {
+        input_ipfs_path: opts.input_ipfs_path,
+    };
     db.event_store.migrate_from_ipfs(network, blocks).await?;
     Ok(())
 }
-fn blocks_from_filesystem(input_ipfs_path: PathBuf) -> impl Stream<Item = Result<BoxedBlock>> {
-    // the block store is split in to 1024 directories and then the blocks stored as files.
-    // the dir structure is the penultimate two characters as dir then the b32 sha256 multihash of the block
-    // The leading "B" for the b32 sha256 multihash is left off
-    // ~/.ipfs/blocks/QV/CIQOHMGEIKMPYHAUTL57JSEZN64SIJ5OIHSGJG4TJSSJLGI3PBJLQVI.data // cspell:disable-line
-    info!(path = %input_ipfs_path.display(), "opening IPFS repo");
-
-    let mut dirs = Vec::new();
-    dirs.push(input_ipfs_path);
-
-    try_stream! {
-        while !dirs.is_empty() {
-            let mut entries = tokio::fs::read_dir(dirs.pop().unwrap()).await?;
-            while let Some(entry) = entries.next_entry().await? {
-                if entry.metadata().await?.is_dir() {
-                    dirs.push(entry.path())
-                } else if let Some(block) = block_from_path(entry.path()).await?{
-                    yield block
+
+struct FSBlockStore {
+    input_ipfs_path: PathBuf,
+}
+
+#[async_trait]
+impl BlockStore for FSBlockStore {
+    fn blocks(&self) -> BoxStream<'static, anyhow::Result<(Cid, Vec<u8>)>> {
+        // the block store is split in to 1024 directories and then the blocks stored as files.
+        // the dir structure is the penultimate two characters as dir then the b32 sha256 multihash of the block
+        // The leading "B" for the b32 sha256 multihash is left off
+        // ~/.ipfs/blocks/QV/CIQOHMGEIKMPYHAUTL57JSEZN64SIJ5OIHSGJG4TJSSJLGI3PBJLQVI.data // cspell:disable-line
+        info!(path = %self.input_ipfs_path.display(), "opening IPFS repo");
+
+        let mut dirs = Vec::new();
+        dirs.push(self.input_ipfs_path.clone());
+
+        try_stream! {
+            while !dirs.is_empty() {
+                let mut entries = tokio::fs::read_dir(dirs.pop().unwrap()).await?;
```

Comment:
nit: I was worried this might crash on an empty dir before I realized the while condition.. my preference is probably `while let Some(dir) = dirs.pop()` but same same.